Add label documentation (#190)
eclipsesource/jsonforms2-website: content/docs/labels.mdx

---
id: labels
title: Labels
---

import { OneOfLabel, PropertyLabel, SchemaLabel, UiSchemaLabel } from '../../src/components/docs/labels'

> There are three different ways, how a label is determined by JSON Forms in the following order:

Reviewer: ```suggestion
Labels are determined by JSON Forms in the following way:

```

- If an UI Schema `label` is set, it will be used as the label
- If there is no UI Schema `label` **but** a JSON Schema `title`, the JSON Schema `title` will be used as the label

> - If there is no UI schema `label` **and** no JSON Schema `title`, the label will be derived derived from the property name

Reviewer: ```suggestion
- If there is no UI schema `label` **and** no JSON Schema `title`, the label will be derived from the property name

```

> ## Derived by property

Reviewer: It would be nice to order the sections in the same way as outlined in the order steps

When no `label` in the UI Schema and no `title` in the JSON Schema is provided, JSON Forms will derive the label directly from the property name.
For example if the input name is `firstname`, the label will be `Firstname`.
In order to have a label with multiple separated words, camel case can be used.
So if the input name is `firstName`, the label will be `First Name`.

<PropertyLabel />

## JSON Schema `title`

> If you want to set a label explicitly, one way is to use the `title` attribute within JSON schema.

Reviewer: ```suggestion
When no UI Schema `label` is set, JSON Forms will use the JSON Schema `title` annotation.

```

```js
{
  properties: {
    name: {
      type: "string",
      title: "First Name"
    }
  }
}
```

<SchemaLabel />

## UI Schema `label`

> Alternatively you can also set a label by using the `label` attribute within the UI schema.

Reviewer: Labels can be specified in UI Schemas via the `label` attribute.
If both the JSON Schema `title` and the UI Schema `label` are specified, the UI Schema `label` will be used.

```js
{
  type: "VerticalLayout",
  elements: [
    {
      type: "Control",
      scope: "#/properties/name",
      label: "First Name"
    }
  ]
}
```

<UiSchemaLabel />

## `oneOf` enum titles

> Since JSON Schema does not provide any way to set custom labels for enums, you can use an `oneOf` enum construct in order to give your enums specific labels:

Reviewer: ```suggestion
JSON Schema does not provide a way to specify titles for enum values.
However you can use a `oneOf` construct instead in which each entry specifies a `const` for the enum value and `title` for the enum label.
JSON Forms will recognize this construct and render it in the same way as usual enums.
```
